src/charm.py

```diff
@@ -702,11 +702,32 @@ def _generate_prometheus_config(self) -> bool:
         scrape_jobs = self.metrics_consumer.jobs()
         for job in scrape_jobs:
             job["honor_labels"] = True
-            if (tls_config := job.get("tls_config")) and (ca_file := tls_config.get("ca_file")):
-                # Cert is transferred over relation data and needs to be written to a file on disk.
-                cert_filename = f"/etc/prometheus/{job['job_name']}.crt"
-                certs[cert_filename] = ca_file
-                job["tls_config"]["ca_file"] = cert_filename
+            if tls_config := job.get("tls_config"):
+                # Certs are transferred over relation data and need to be written to files on disk.
+                # CA certificate to validate the server certificate with.
+                if ca_file := tls_config.get("ca_file"):
+                    filename = f"{PROMETHEUS_DIR}/{job['job_name']}.crt"
+                    certs[filename] = ca_file
+                    job["tls_config"]["ca_file"] = filename
+                # Certificate and key files for client cert authentication to the server.
+                if (cert_file := tls_config.get("cert_file")) and (
+                    key_file := tls_config.get("key_file")
+                ):
+                    filename = f"{PROMETHEUS_DIR}/client-{job['job_name']}.crt"
+                    certs[filename] = cert_file
+                    job["tls_config"]["cert_file"] = filename
+                    filename = f"{PROMETHEUS_DIR}/client-{job['job_name']}.key"
+                    certs[filename] = key_file
+                    job["tls_config"]["key_file"] = filename
+                elif "cert_file" in tls_config or "key_file" in tls_config:
+                    logger.error(
+                        'tls_config requires both "cert_file" and "key_file" if client authentication is to be used'
+                    )
+                    self.unit.status = BlockedStatus(
+                        "Invalid prometheus configuration; see debug logs"
+                    )
+                    return False
```

Comment:
**issue**: Prometheus sets status in a centralized way - we shouldn't set it here.

**issue**: This functions "Returns a boolean indicating if a new configuration was pushed", so `return False` is not appropriate here.

**suggestion**: Raise a new exception here, e.g. `ConfigError`, and catch it in `_configure`:

```diff
        try:
            # Need to reload if config or alerts changed.
            # (Both functions need to run so cannot use the short-circuiting `or`.)
            should_reload = any(
                [
                    self._generate_prometheus_config(),
                    self._set_alerts(),
                ]
            )
+       except ConfigError as e:
+           logger.error("Failed to generate configuration: %s", e)
+           self.unit.status = str(e)
+           return
        except PebbleError as e:
```